// @formily/antd-v5/esm/editable/style.js
import { genStyleHook } from './../__builtins__';
export default genStyleHook('editable', function (token) {
    var _a, _b, _c;
    var antCls = token.antCls, componentCls = token.componentCls, fontSizeSM = token.fontSizeSM, colorPrimary = token.colorPrimary;
    return _a = {},
        _a[componentCls] = (_b = {
                cursor: 'pointer',
                display: 'inline-block !important'
            },
            _b["".concat(antCls, "-form-text")] = (_c = {
                    display: 'flex',
                    alignItems: 'center'
                },
                _c["".concat(antCls, "-tag")] = {
                    transition: 'none !important',
                },
                _c["".concat(antCls, "-tag:last-child")] = {
                    paddingInlineEnd: '0 !important',
                },
                _c),
            _b["".concat(componentCls, "-content")] = {
                display: 'flex',
                alignItems: 'center',
                '> *': {
                    marginInlineEnd: 3,
                    '&:last-child': {
                        marginInlineEnd: 0,
                    },
                },
            },
            _b["".concat(componentCls, "-edit-btn,\n        ").concat(componentCls, "-close-btn")] = {
                transition: 'all .25s ease-in-out',
                color: '#aaa',
                fontSize: fontSizeSM,
                '&:hover': {
                    color: colorPrimary,
                },
            },
            _b["".concat(componentCls, "-preview")] = {
                whiteSpace: 'nowrap',
                textOverflow: 'ellipsis',
                overflow: 'hidden',
                wordBreak: 'break-all',
                maxWidth: '100%',
                display: 'block',
            },
            _b),
        _a;
});
