// @formily/antd-v5/lib/array-table/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArrayTable = void 0;
var react_1 = require("@formily/react");
var shared_1 = require("@formily/shared");
var antd_1 = require("antd");
var classnames_1 = __importDefault(require("classnames"));
var react_2 = __importStar(require("react"));
var array_base_1 = require("../array-base");
var __builtins__1 = require("../__builtins__");
var style_1 = __importDefault(require("./style"));
var SortableRow = (0, __builtins__1.SortableElement)(function (props) { return react_2.default.createElement("tr", __assign({}, props)); });
var SortableBodyRaw = (0, __builtins__1.SortableContainer)(function (_a) {
    var tbodyRef = _a.tbodyRef, props = __rest(_a, ["tbodyRef"]);
    return react_2.default.createElement("tbody", __assign({}, props, { ref: tbodyRef }));
});
var SortableBody = (0, react_2.forwardRef)(function (props, ref) { return react_2.default.createElement(SortableBodyRaw, __assign({}, props, { tbodyRef: ref })); });
var isColumnComponent = function (schema) {
    var _a;
    return ((_a = schema['x-component']) === null || _a === void 0 ? void 0 : _a.indexOf('Column')) > -1;
};
var isOperationsComponent = function (schema) {
    var _a;
    return ((_a = schema['x-component']) === null || _a === void 0 ? void 0 : _a.indexOf('Operations')) > -1;
};
var isAdditionComponent = function (schema) {
    var _a;
    return ((_a = schema['x-component']) === null || _a === void 0 ? void 0 : _a.indexOf('Addition')) > -1;
};
var useArrayTableSources = function () {
    var arrayField = (0, react_1.useField)();
    var schema = (0, react_1.useFieldSchema)();
    var parseSources = function (schema) {
        var _a, _b, _c;
        if (isColumnComponent(schema) ||
            isOperationsComponent(schema) ||
            isAdditionComponent(schema)) {
            if (!((_a = schema['x-component-props']) === null || _a === void 0 ? void 0 : _a['dataIndex']) && !schema['name'])
                return [];
            var name_1 = ((_b = schema['x-component-props']) === null || _b === void 0 ? void 0 : _b['dataIndex']) || schema['name'];
            var field = arrayField.query(arrayField.address.concat(name_1)).take();
            var columnProps = ((_c = field === null || field === void 0 ? void 0 : field.component) === null || _c === void 0 ? void 0 : _c[1]) || schema['x-component-props'] || {};
            var display = (field === null || field === void 0 ? void 0 : field.display) || schema['x-display'];
            return [
                {
                    name: name_1,
                    display: display,
                    field: field,
                    schema: schema,
                    columnProps: columnProps,
                },
            ];
        }
        else if (schema.properties) {
            return schema.reduceProperties(function (buf, schema) {
                return buf.concat(parseSources(schema));
            }, []);
        }
        return [];
    };
    var parseArrayItems = function (schema) {
        if (!schema)
            return [];
        var sources = [];
        var items = (0, shared_1.isArr)(schema) ? schema : [schema];
        return items.reduce(function (columns, schema) {
            var item = parseSources(schema);
            if (item) {
                return columns.concat(item);
            }
            return columns;
        }, sources);
    };
    if (!schema)
        throw new Error('can not found schema object');
    return parseArrayItems(schema.items);
};
var indexKey = '__DO_NOT_USE_THIS_PROPERTY_index__';
var useArrayTableColumns = function (dataSource, field, sources) {
    return sources.reduce(function (buf, _a, key) {
        var name = _a.name, columnProps = _a.columnProps, schema = _a.schema, display = _a.display;
        if (display !== 'visible')
            return buf;
        if (!isColumnComponent(schema))
            return buf;
        return buf.concat(__assign(__assign({}, columnProps), { key: key, dataIndex: name, render: function (value, record) {
                var index = record[indexKey];
                var children = (react_2.default.createElement(array_base_1.ArrayBase.Item, { index: index, record: function () { var _a; return (_a = field === null || field === void 0 ? void 0 : field.value) === null || _a === void 0 ? void 0 : _a[index]; } },
                    react_2.default.createElement(react_1.RecursionField, { schema: schema, name: index, onlyRenderProperties: true })));
                return index > -1 ? children : null;
            } }));
    }, []);
};
var useAddition = function () {
    var schema = (0, react_1.useFieldSchema)();
    return schema.reduceProperties(function (addition, schema, key) {
        if (isAdditionComponent(schema)) {
            return react_2.default.createElement(react_1.RecursionField, { schema: schema, name: key });
        }
        return addition;
    }, null);
};
var schedulerRequest = {
    request: null,
};
var StatusSelect = (0, react_1.observer)(function (props) {
    var _a;
    var field = (0, react_1.useField)();
    var prefixCls = (0, __builtins__1.usePrefixCls)('formily-array-table');
    var _b = __read((0, style_1.default)(prefixCls), 2), wrapSSR = _b[0], hashId = _b[1];
    var errors = field.errors;
    var parseIndex = function (address) {
        var _a;
        return Number((_a = address === null || address === void 0 ? void 0 : address.slice(address.indexOf(field.address.toString())).match(/(\d+)/)) === null || _a === void 0 ? void 0 : _a[1]);
    };
    var options = (_a = props.options) === null || _a === void 0 ? void 0 : _a.map(function (_a) {
        var label = _a.label, value = _a.value;
        var val = Number(value);
        var hasError = errors.some(function (_a) {
            var _b;
            var address = _a.address;
            var currentIndex = parseIndex(address);
            var startIndex = props.pageSize ? (val - 1) * props.pageSize : 0;
            var endIndex = props.pageSize
                ? val * props.pageSize
                : ((_b = props.options) === null || _b === void 0 ? void 0 : _b.length) || 0;
            return currentIndex >= startIndex && currentIndex <= endIndex;
        });
        return {
            label: hasError ? react_2.default.createElement(antd_1.Badge, { dot: true }, label) : label,
            value: value,
        };
    });
    var width = String(options === null || options === void 0 ? void 0 : options.length).length * 15;
    return wrapSSR(react_2.default.createElement(antd_1.Select, { value: props.value, onChange: props.onChange, options: options, virtual: true, style: {
            width: width < 60 ? 60 : width,
        }, className: (0, classnames_1.default)("".concat(prefixCls, "-status-select"), hashId, {
            'has-error': errors === null || errors === void 0 ? void 0 : errors.length,
        }) }));
}, {
    scheduler: function (update) {
        clearTimeout(schedulerRequest.request);
        schedulerRequest.request = setTimeout(function () {
            update();
        }, 100);
    },
});
var PaginationContext = (0, react_2.createContext)({});
var usePagination = function () {
    return (0, react_2.useContext)(PaginationContext);
};
var ArrayTablePagination = function (props) {
    var _a;
    var _b = __read((0, react_2.useState)(1), 2), current = _b[0], setCurrent = _b[1];
    var prefixCls = (0, __builtins__1.usePrefixCls)('formily-array-table');
    var _c = __read((0, style_1.default)(prefixCls), 2), wrapSSR = _c[0], hashId = _c[1];
    var _d = __read((0, react_2.useState)(props.pageSize || 10), 2), pageSize = _d[0], setPageSize = _d[1];
    var size = props.size || 'default';
    var dataSource = props.dataSource || [];
    var showPagination = props.showPagination;
    var startIndex = (current - 1) * pageSize;
    var endIndex = startIndex + pageSize - 1;
    var total = (dataSource === null || dataSource === void 0 ? void 0 : dataSource.length) || 0;
    var totalPage = Math.ceil(total / pageSize);
    var pages = Array.from(new Array(totalPage)).map(function (_, index) {
        var page = index + 1;
        return {
            label: page,
            value: page,
        };
    });
    var handleChange = function (current) {
        setCurrent(current);
    };
    var handleSizeChange = function (_, size) {
        setPageSize(size);
    };
    (0, react_2.useEffect)(function () {
        setPageSize(props.pageSize || 10);
    }, [props.pageSize]);
    (0, react_2.useEffect)(function () {
        if (totalPage > 0 && totalPage < current) {
            handleChange(totalPage);
        }
    }, [totalPage, current]);
    var renderPagination = function () {
        if (!showPagination || totalPage < 1)
            return;
        return (react_2.default.createElement("div", { className: (0, classnames_1.default)("".concat(prefixCls, "-pagination"), hashId) },
            react_2.default.createElement(antd_1.Space, null,
                react_2.default.createElement(StatusSelect, { value: current, pageSize: pageSize, onChange: handleChange, options: pages, notFoundContent: false }),
                react_2.default.createElement(antd_1.Pagination, __assign({}, props, { pageSize: pageSize, current: current, total: dataSource.length, size: size, showSizeChanger: true, onShowSizeChange: handleSizeChange, onChange: handleChange })))));
    };
    return wrapSSR(react_2.default.createElement(react_2.Fragment, null,
        react_2.default.createElement(PaginationContext.Provider, { value: { totalPage: totalPage, pageSize: pageSize, startIndex: startIndex, changePage: handleChange } }, (_a = props.children) === null || _a === void 0 ? void 0 : _a.call(props, showPagination
            ? dataSource === null || dataSource === void 0 ? void 0 : dataSource.slice(startIndex, endIndex + 1)
            : dataSource, renderPagination(), {
            startIndex: startIndex,
        }))));
};
var WrapperComp = function (props) {
    var ref = (0, react_2.useRef)(null);
    var field = (0, react_1.useField)();
    var prefixCls = (0, __builtins__1.usePrefixCls)('formily-array-table');
    var startIndex = (0, react_2.useContext)(PaginationContext).startIndex;
    var dataSource = Array.isArray(field.value) ? field.value.slice() : [];
    var addTdStyles = function (id) {
        var _a;
        var node = (_a = ref.current) === null || _a === void 0 ? void 0 : _a.querySelector(".".concat(prefixCls, "-row-").concat(id));
        var helper = document.body.querySelector(".".concat(prefixCls, "-sort-helper"));
        if (!helper)
            return;
        var tds = node === null || node === void 0 ? void 0 : node.querySelectorAll('td');
        if (!tds)
            return;
        requestAnimationFrame(function () {
            helper.querySelectorAll('td').forEach(function (td, index) {
                if (tds[index]) {
                    td.style.width = getComputedStyle(tds[index]).width;
                }
            });
        });
    };
    return (react_2.default.createElement(SortableBody, __assign({ ref: ref }, props, { start: startIndex, list: dataSource.slice(), accessibility: {
            container: ref.current || undefined,
        }, onSortStart: function (event) {
            addTdStyles(event.active.id);
        }, onSortEnd: function (_a) {
            var oldIndex = _a.oldIndex, newIndex = _a.newIndex;
            field.move(oldIndex, newIndex);
        }, className: (0, classnames_1.default)("".concat(prefixCls, "-sort-helper"), props.className) })));
};
var RowComp = function (props) {
    var prefixCls = (0, __builtins__1.usePrefixCls)('formily-array-table');
    var index = props['data-row-key'] || 0;
    return (react_2.default.createElement(SortableRow, __assign({ lockAxis: "y" }, props, { index: index, className: (0, classnames_1.default)(props.className, "".concat(prefixCls, "-row-").concat(index + 1)) })));
};
var InternalArrayTable = (0, react_1.observer)(function (props) {
    var ref = (0, react_2.useRef)(null);
    var field = (0, react_1.useField)();
    var prefixCls = (0, __builtins__1.usePrefixCls)('formily-array-table');
    var _a = __read((0, style_1.default)(prefixCls), 2), wrapSSR = _a[0], hashId = _a[1];
    var dataSource = Array.isArray(field.value) ? field.value.slice() : [];
    dataSource.forEach(function (item, index) {
        item[indexKey] = index;
    });
    var sources = useArrayTableSources();
    var columns = useArrayTableColumns(dataSource, field, sources);
    var pagination = (0, shared_1.isBool)(props.pagination) ? {} : props.pagination;
    var showPagination = (0, shared_1.isBool)(props.pagination) ? props.pagination : true;
    var addition = useAddition();
    var defaultRowKey = function (record) {
        return record[indexKey];
    };
    return wrapSSR(react_2.default.createElement(ArrayTablePagination, __assign({}, pagination, { dataSource: dataSource, showPagination: showPagination }), function (dataSource, pager, _a) {
        var startIndex = _a.startIndex;
        return (react_2.default.createElement("div", { ref: ref, className: (0, classnames_1.default)(prefixCls, hashId) },
            react_2.default.createElement(array_base_1.ArrayBase, null,
                react_2.default.createElement(antd_1.Table, __assign({ size: "small", bordered: true, rowKey: defaultRowKey }, props, { onChange: function () { }, pagination: false, columns: columns, dataSource: dataSource, components: {
                        body: {
                            wrapper: WrapperComp,
                            row: RowComp,
                        },
                    } })),
                react_2.default.createElement("div", { style: { marginTop: 5, marginBottom: 5 } }, pager),
                sources.map(function (column, key) {
                    //专门用来承接对Column的状态管理
                    if (!isColumnComponent(column.schema))
                        return;
                    return react_2.default.createElement(react_1.RecursionField, {
                        name: column.name,
                        schema: column.schema,
                        onlyRenderSelf: true,
                        key: key,
                    });
                }),
                addition)));
    }));
});
var Column = function () {
    return react_2.default.createElement(react_2.Fragment, null);
};
var Addition = function (props) {
    var array = array_base_1.ArrayBase.useArray();
    var _a = usePagination(), _b = _a.totalPage, totalPage = _b === void 0 ? 0 : _b, _c = _a.pageSize, pageSize = _c === void 0 ? 10 : _c, changePage = _a.changePage;
    return (react_2.default.createElement(array_base_1.ArrayBase.Addition, __assign({}, props, { onClick: function (e) {
            var _a, _b;
            // 如果添加数据后将超过当前页，则自动切换到下一页
            var total = ((_a = array === null || array === void 0 ? void 0 : array.field) === null || _a === void 0 ? void 0 : _a.value.length) || 0;
            if (total === totalPage * pageSize + 1 && (0, shared_1.isFn)(changePage)) {
                changePage(totalPage + 1);
            }
            (_b = props.onClick) === null || _b === void 0 ? void 0 : _b.call(props, e);
        } })));
};
exports.ArrayTable = Object.assign(array_base_1.ArrayBase.mixin(InternalArrayTable), {
    Column: Column,
    Addition: Addition,
});
exports.ArrayTable.displayName = 'ArrayTable';
exports.default = exports.ArrayTable;
