// @formily/antd-v5/lib/__builtins__/style.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.genStyleHook = exports.genCommonStyle = void 0;
var cssinjs_1 = require("@ant-design/cssinjs");
var shared_1 = require("@formily/shared");
var hooks_1 = require("./hooks");
var genCommonStyle = function (token, componentPrefixCls) {
    var _a, _b;
    var fontFamily = token.fontFamily, fontSize = token.fontSize;
    var rootPrefixSelector = "[class^=\"".concat(componentPrefixCls, "\"], [class*=\" ").concat(componentPrefixCls, "\"]");
    return _a = {},
        _a[rootPrefixSelector] = (_b = {
                fontFamily: fontFamily,
                fontSize: fontSize,
                boxSizing: 'border-box',
                '&::before, &::after': {
                    boxSizing: 'border-box',
                }
            },
            _b[rootPrefixSelector] = {
                boxSizing: 'border-box',
                '&::before, &::after': {
                    boxSizing: 'border-box',
                },
            },
            _b),
        _a;
};
exports.genCommonStyle = genCommonStyle;
var genStyleHook = function (component, styleFn) {
    return function (prefixCls) {
        var _a = (0, hooks_1.useToken)(), theme = _a.theme, token = _a.token, hashId = _a.hashId;
        var _b = (0, hooks_1.useConfig)(), getPrefixCls = _b.getPrefixCls, iconPrefixCls = _b.iconPrefixCls, csp = _b.csp;
        var rootPrefixCls = getPrefixCls();
        return [
            (0, cssinjs_1.useStyleRegister)({
                nonce: csp === null || csp === void 0 ? void 0 : csp.nonce,
                theme: theme,
                token: token,
                hashId: hashId,
                path: ['formily-antd', component, prefixCls, iconPrefixCls],
            }, function () {
                var componentCls = ".".concat(prefixCls);
                var mergedToken = (0, shared_1.merge)(token, __assign(__assign({}, token['Form']), { // Merge the antd form token
                    componentCls: componentCls, prefixCls: prefixCls, iconCls: ".".concat(iconPrefixCls), antCls: ".".concat(rootPrefixCls) }));
                var styleInterpolation = styleFn(mergedToken, {
                    hashId: hashId,
                    prefixCls: prefixCls,
                    rootPrefixCls: rootPrefixCls,
                    iconPrefixCls: iconPrefixCls,
                });
                return [(0, exports.genCommonStyle)(token, prefixCls), styleInterpolation];
            }),
            hashId,
        ];
    };
};
exports.genStyleHook = genStyleHook;
