// @formily/antd-v5/lib/form-item/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormItem = exports.BaseItem = void 0;
var icons_1 = require("@ant-design/icons");
var core_1 = require("@formily/core");
var react_1 = require("@formily/react");
var antd_1 = require("antd");
var classnames_1 = __importDefault(require("classnames"));
var react_2 = __importStar(require("react"));
var form_layout_1 = require("../form-layout");
var __builtins__1 = require("../__builtins__");
var hooks_1 = require("./hooks");
var style_1 = __importDefault(require("./style"));
__exportStar(require("./hooks"), exports);
__exportStar(require("./types"), exports);
var ICON_MAP = {
    error: react_2.default.createElement(icons_1.CloseCircleOutlined, null),
    success: react_2.default.createElement(icons_1.CheckCircleOutlined, null),
    warning: react_2.default.createElement(icons_1.ExclamationCircleOutlined, null),
};
var BaseItem = function (_a) {
    var _b, _c, _d, _e, _f;
    var children = _a.children, props = __rest(_a, ["children"]);
    var _g = __read((0, react_2.useState)(false), 2), active = _g[0], setActive = _g[1];
    var formLayout = (0, hooks_1.useFormItemLayout)(props);
    var _h = (0, hooks_1.useOverflow)(), containerRef = _h.containerRef, contentRef = _h.contentRef, overflow = _h.overflow;
    var label = formLayout.label, style = formLayout.style, layout = formLayout.layout, _j = formLayout.colon, colon = _j === void 0 ? true : _j, addonBefore = formLayout.addonBefore, addonAfter = formLayout.addonAfter, asterisk = formLayout.asterisk, feedbackStatus = formLayout.feedbackStatus, extra = formLayout.extra, feedbackText = formLayout.feedbackText, fullness = formLayout.fullness, feedbackLayout = formLayout.feedbackLayout, feedbackIcon = formLayout.feedbackIcon, getPopupContainer = formLayout.getPopupContainer, inset = formLayout.inset, _k = formLayout.bordered, bordered = _k === void 0 ? true : _k, labelWidth = formLayout.labelWidth, wrapperWidth = formLayout.wrapperWidth, labelCol = formLayout.labelCol, wrapperCol = formLayout.wrapperCol, labelAlign = formLayout.labelAlign, _l = formLayout.wrapperAlign, wrapperAlign = _l === void 0 ? 'left' : _l, size = formLayout.size, labelWrap = formLayout.labelWrap, wrapperWrap = formLayout.wrapperWrap, tooltipLayout = formLayout.tooltipLayout, tooltip = formLayout.tooltip, tooltipIcon = formLayout.tooltipIcon;
    var labelStyle = __assign({}, formLayout.labelStyle);
    var wrapperStyle = __assign({}, formLayout.wrapperStyle);
    var prefixCls = (0, __builtins__1.usePrefixCls)('formily-item', props);
    var _m = __read((0, style_1.default)(prefixCls), 2), wrapSSR = _m[0], hashId = _m[1];
    // 固定宽度
    var enableCol = false;
    if (labelWidth || wrapperWidth) {
        if (labelWidth) {
            labelStyle.width = labelWidth === 'auto' ? undefined : labelWidth;
            labelStyle.maxWidth = labelWidth === 'auto' ? undefined : labelWidth;
        }
        if (wrapperWidth) {
            wrapperStyle.width = wrapperWidth === 'auto' ? undefined : wrapperWidth;
            wrapperStyle.maxWidth = wrapperWidth === 'auto' ? undefined : wrapperWidth;
        }
        // 栅格模式
    }
    if (labelCol || wrapperCol) {
        if (!labelStyle.width && !wrapperStyle.width && layout !== 'vertical') {
            enableCol = true;
        }
    }
    var formatChildren = feedbackLayout === 'popover' ? (react_2.default.createElement(antd_1.Popover, { autoAdjustOverflow: true, overlayClassName: "".concat(prefixCls, "-popover"), placement: "top", content: react_2.default.createElement("div", { className: (0, classnames_1.default)((_b = {},
                _b["".concat(prefixCls, "-").concat(feedbackStatus, "-help")] = !!feedbackStatus,
                _b["".concat(prefixCls, "-help")] = true,
                _b)) },
            feedbackStatus ? ICON_MAP[feedbackStatus] : null,
            " ",
            feedbackText), open: !!feedbackText, getPopupContainer: getPopupContainer },
        react_2.default.createElement(react_2.default.Fragment, null, children))) : (children);
    var gridStyles = {};
    var getOverflowTooltip = function () {
        if (overflow) {
            return (react_2.default.createElement("div", null,
                react_2.default.createElement("div", null, label),
                react_2.default.createElement("div", null, tooltip)));
        }
        return tooltip;
    };
    var renderLabelText = function () {
        var labelChildren = (react_2.default.createElement("div", { className: "".concat(prefixCls, "-label-content"), ref: containerRef },
            react_2.default.createElement("span", { ref: contentRef },
                asterisk && react_2.default.createElement("span", { className: "".concat(prefixCls, "-asterisk") }, '*'),
                react_2.default.createElement("label", null, label))));
        if ((tooltipLayout === 'text' && tooltip) || overflow) {
            return (react_2.default.createElement(antd_1.Tooltip, { placement: "top", title: getOverflowTooltip() }, labelChildren));
        }
        return labelChildren;
    };
    var renderTooltipIcon = function () {
        if (tooltip && tooltipLayout === 'icon' && !overflow) {
            return (react_2.default.createElement("span", { className: "".concat(prefixCls, "-label-tooltip-icon") },
                react_2.default.createElement(antd_1.Tooltip, { placement: "top", title: tooltip }, tooltipIcon)));
        }
    };
    var renderLabel = function () {
        var _a;
        if (!label)
            return null;
        return (react_2.default.createElement("div", { className: (0, classnames_1.default)((_a = {},
                _a["".concat(prefixCls, "-label")] = true,
                _a["".concat(prefixCls, "-label-tooltip")] = (tooltip && tooltipLayout === 'text') || overflow,
                _a["".concat(prefixCls, "-item-col-").concat(labelCol)] = enableCol && !!labelCol,
                _a)), style: labelStyle },
            renderLabelText(),
            renderTooltipIcon(),
            label !== ' ' && (react_2.default.createElement("span", { className: "".concat(prefixCls, "-colon") }, colon ? ':' : ''))));
    };
    return wrapSSR(react_2.default.createElement("div", __assign({}, (0, __builtins__1.pickDataProps)(props), { style: __assign(__assign({}, style), gridStyles), "data-grid-span": props.gridSpan, className: (0, classnames_1.default)(props.className, (_c = {},
            _c["".concat(prefixCls)] = true,
            _c["".concat(prefixCls, "-layout-").concat(layout)] = true,
            _c["".concat(prefixCls, "-").concat(feedbackStatus)] = !!feedbackStatus,
            _c["".concat(prefixCls, "-feedback-has-text")] = !!feedbackText,
            _c["".concat(prefixCls, "-size-").concat(size)] = !!size,
            _c["".concat(prefixCls, "-feedback-layout-").concat(feedbackLayout)] = !!feedbackLayout,
            _c["".concat(prefixCls, "-fullness")] = !!fullness || !!inset || !!feedbackIcon,
            _c["".concat(prefixCls, "-inset")] = !!inset,
            _c["".concat(prefixCls, "-active")] = active,
            _c["".concat(prefixCls, "-inset-active")] = !!inset && active,
            _c["".concat(prefixCls, "-label-align-").concat(labelAlign)] = true,
            _c["".concat(prefixCls, "-control-align-").concat(wrapperAlign)] = true,
            _c["".concat(prefixCls, "-label-wrap")] = !!labelWrap,
            _c["".concat(prefixCls, "-control-wrap")] = !!wrapperWrap,
            _c["".concat(prefixCls, "-bordered-none")] = bordered === false || !!inset || !!feedbackIcon,
            _c), hashId), onFocus: function () {
            if (feedbackIcon || inset) {
                setActive(true);
            }
        }, onBlur: function () {
            if (feedbackIcon || inset) {
                setActive(false);
            }
        } }),
        renderLabel(),
        react_2.default.createElement("div", { className: (0, classnames_1.default)((_d = {},
                _d["".concat(prefixCls, "-control")] = true,
                _d["".concat(prefixCls, "-item-col-").concat(wrapperCol)] = enableCol && !!wrapperCol && label,
                _d)) },
            react_2.default.createElement("div", { className: (0, classnames_1.default)("".concat(prefixCls, "-control-content")) },
                addonBefore && (react_2.default.createElement("div", { className: (0, classnames_1.default)("".concat(prefixCls, "-addon-before")) }, addonBefore)),
                react_2.default.createElement("div", { style: wrapperStyle, className: (0, classnames_1.default)((_e = {},
                        _e["".concat(prefixCls, "-control-content-component")] = true,
                        _e["".concat(prefixCls, "-control-content-component-has-feedback-icon")] = !!feedbackIcon,
                        _e)) },
                    react_2.default.createElement(form_layout_1.FormLayoutShallowContext.Provider, { value: {} }, formatChildren),
                    feedbackIcon && (react_2.default.createElement("div", { className: (0, classnames_1.default)("".concat(prefixCls, "-feedback-icon")) }, feedbackIcon))),
                addonAfter && (react_2.default.createElement("div", { className: (0, classnames_1.default)("".concat(prefixCls, "-addon-after")) }, addonAfter))),
            !!feedbackText &&
                feedbackLayout !== 'popover' &&
                feedbackLayout !== 'none' && (react_2.default.createElement("div", { className: (0, classnames_1.default)((_f = {},
                    _f["".concat(prefixCls, "-").concat(feedbackStatus, "-help")] = !!feedbackStatus,
                    _f["".concat(prefixCls, "-help")] = true,
                    _f["".concat(prefixCls, "-help-enter")] = true,
                    _f["".concat(prefixCls, "-help-enter-active")] = true,
                    _f)) }, feedbackText)),
            extra && react_2.default.createElement("div", { className: (0, classnames_1.default)("".concat(prefixCls, "-extra")) }, extra))));
};
exports.BaseItem = BaseItem;
// 适配
exports.FormItem = Object.assign((0, react_1.connect)(exports.BaseItem, (0, react_1.mapProps)(function (props, field) {
    if ((0, core_1.isVoidField)(field))
        return {
            label: field.title || props.label,
            asterisk: props.asterisk,
            extra: props.extra || field.description,
        };
    if (!field)
        return props;
    var takeFeedbackStatus = function () {
        if (field.validating)
            return 'pending';
        return field.decoratorProps.feedbackStatus || field.validateStatus;
    };
    var takeMessage = function () {
        var split = function (messages) {
            return messages.reduce(function (buf, text, index) {
                if (!text)
                    return buf;
                return index < messages.length - 1
                    ? buf.concat([text, ', '])
                    : buf.concat([text]);
            }, []);
        };
        if (field.validating)
            return;
        if (props.feedbackText)
            return props.feedbackText;
        if (field.selfErrors.length)
            return split(field.selfErrors);
        if (field.selfWarnings.length)
            return split(field.selfWarnings);
        if (field.selfSuccesses.length)
            return split(field.selfSuccesses);
    };
    var takeAsterisk = function () {
        if ('asterisk' in props) {
            return props.asterisk;
        }
        if (field.required && field.pattern !== 'readPretty') {
            return true;
        }
        return false;
    };
    return {
        label: props.label || field.title,
        feedbackStatus: takeFeedbackStatus(),
        feedbackText: takeMessage(),
        asterisk: takeAsterisk(),
        extra: props.extra || field.description,
    };
})), {
    BaseItem: exports.BaseItem,
});
exports.default = exports.FormItem;
