// @formily/antd-v5/lib/array-base/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArrayBase = void 0;
var icons_1 = require("@ant-design/icons");
var react_1 = require("@formily/react");
var shared_1 = require("@formily/shared");
var antd_1 = require("antd");
var classnames_1 = __importDefault(require("classnames"));
var react_2 = __importStar(require("react"));
var __builtins__1 = require("../__builtins__");
var style_1 = __importDefault(require("./style"));
var ArrayBaseContext = (0, react_2.createContext)(null);
var ItemContext = (0, react_2.createContext)(null);
var takeRecord = function (val, index) {
    return typeof val === 'function' ? val(index) : val;
};
var useArray = function () {
    return (0, react_2.useContext)(ArrayBaseContext);
};
var useIndex = function (index) {
    var ctx = (0, react_2.useContext)(ItemContext);
    return (ctx ? ctx.index : index) || 0;
};
var useRecord = function (record) {
    var ctx = (0, react_2.useContext)(ItemContext);
    return takeRecord(ctx ? ctx.record : record, ctx === null || ctx === void 0 ? void 0 : ctx.index);
};
var getSchemaDefaultValue = function (schema) {
    if ((schema === null || schema === void 0 ? void 0 : schema.type) === 'array')
        return [];
    if ((schema === null || schema === void 0 ? void 0 : schema.type) === 'object')
        return {};
    if ((schema === null || schema === void 0 ? void 0 : schema.type) === 'void') {
        for (var key in schema.properties) {
            var value = getSchemaDefaultValue(schema.properties[key]);
            if ((0, shared_1.isValid)(value))
                return value;
        }
    }
};
var getDefaultValue = function (defaultValue, schema) {
    if ((0, shared_1.isValid)(defaultValue))
        return (0, shared_1.clone)(defaultValue);
    if (Array.isArray(schema === null || schema === void 0 ? void 0 : schema.items))
        return getSchemaDefaultValue(schema === null || schema === void 0 ? void 0 : schema.items[0]);
    return getSchemaDefaultValue(schema === null || schema === void 0 ? void 0 : schema.items);
};
var InternalArrayBase = function (props) {
    var field = (0, react_1.useField)();
    var schema = (0, react_1.useFieldSchema)();
    return (react_2.default.createElement(react_1.RecordsScope, { getRecords: function () { return field.value; } },
        react_2.default.createElement(ArrayBaseContext.Provider, { value: { field: field, schema: schema, props: props } }, props.children)));
};
var Item = function (_a) {
    var children = _a.children, props = __rest(_a, ["children"]);
    return (react_2.default.createElement(ItemContext.Provider, { value: props },
        react_2.default.createElement(react_1.RecordScope, { getIndex: function () { return props.index; }, getRecord: function () { return takeRecord(props.record, props.index); } }, children)));
};
var InternalSortHandle = (0, __builtins__1.SortableHandle)(function (props) {
    var prefixCls = (0, __builtins__1.usePrefixCls)('formily-array-base');
    var _a = __read((0, style_1.default)(prefixCls), 2), wrapSSR = _a[0], hashId = _a[1];
    return wrapSSR(react_2.default.createElement(icons_1.MenuOutlined, __assign({}, props, { className: (0, classnames_1.default)("".concat(prefixCls, "-sort-handle"), hashId, props.className), style: __assign({}, props.style) })));
});
var SortHandle = function (props) {
    var _a;
    var array = useArray();
    if (!array)
        return null;
    if (((_a = array.field) === null || _a === void 0 ? void 0 : _a.pattern) !== 'editable')
        return null;
    return react_2.default.createElement(InternalSortHandle, __assign({}, props));
};
var Index = function (props) {
    var index = useIndex();
    var prefixCls = (0, __builtins__1.usePrefixCls)('formily-array-base');
    var _a = __read((0, style_1.default)(prefixCls), 2), wrapSSR = _a[0], hashId = _a[1];
    return wrapSSR(react_2.default.createElement("span", __assign({}, props, { className: (0, classnames_1.default)("".concat(prefixCls, "-index"), hashId) }),
        "#",
        (index || 0) + 1,
        "."));
};
var Addition = function (props) {
    var _a, _b;
    var self = (0, react_1.useField)();
    var array = useArray();
    var prefixCls = (0, __builtins__1.usePrefixCls)('formily-array-base');
    var _c = __read((0, style_1.default)(prefixCls), 2), wrapSSR = _c[0], hashId = _c[1];
    if (!array)
        return null;
    if (((_a = array.field) === null || _a === void 0 ? void 0 : _a.pattern) !== 'editable' &&
        ((_b = array.field) === null || _b === void 0 ? void 0 : _b.pattern) !== 'disabled')
        return null;
    return wrapSSR(react_2.default.createElement(antd_1.Button, __assign({ type: "dashed", block: true }, props, { disabled: self === null || self === void 0 ? void 0 : self.disabled, className: (0, classnames_1.default)("".concat(prefixCls, "-addition"), hashId, props.className), onClick: function (e) {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
            if ((_a = array.props) === null || _a === void 0 ? void 0 : _a.disabled)
                return;
            var defaultValue = getDefaultValue(props.defaultValue, array.schema);
            if (props.method === 'unshift') {
                (_c = (_b = array.field) === null || _b === void 0 ? void 0 : _b.unshift) === null || _c === void 0 ? void 0 : _c.call(_b, defaultValue);
                (_e = (_d = array.props) === null || _d === void 0 ? void 0 : _d.onAdd) === null || _e === void 0 ? void 0 : _e.call(_d, 0);
            }
            else {
                (_g = (_f = array.field) === null || _f === void 0 ? void 0 : _f.push) === null || _g === void 0 ? void 0 : _g.call(_f, defaultValue);
                (_j = (_h = array.props) === null || _h === void 0 ? void 0 : _h.onAdd) === null || _j === void 0 ? void 0 : _j.call(_h, ((_l = (_k = array === null || array === void 0 ? void 0 : array.field) === null || _k === void 0 ? void 0 : _k.value) === null || _l === void 0 ? void 0 : _l.length) - 1);
            }
            if (props.onClick) {
                props.onClick(e);
            }
        }, icon: (0, shared_1.isUndef)(props.icon) ? react_2.default.createElement(icons_1.PlusOutlined, null) : props.icon }), props.title || self.title));
};
var Copy = (0, react_2.forwardRef)(function (props, ref) {
    var _a;
    var self = (0, react_1.useField)();
    var array = useArray();
    var index = useIndex(props.index) || 0;
    var prefixCls = (0, __builtins__1.usePrefixCls)('formily-array-base');
    var _b = __read((0, style_1.default)(prefixCls), 2), wrapSSR = _b[0], hashId = _b[1];
    if (!array)
        return null;
    if (((_a = array.field) === null || _a === void 0 ? void 0 : _a.pattern) !== 'editable')
        return null;
    return wrapSSR(react_2.default.createElement(antd_1.Button, __assign({ type: "ghost" }, props, { style: __assign({ padding: '0 0 0 6px', width: 'auto', height: 'auto' }, props.style), disabled: self === null || self === void 0 ? void 0 : self.disabled, className: (0, classnames_1.default)("".concat(prefixCls, "-copy"), hashId, (self === null || self === void 0 ? void 0 : self.disabled) ? "".concat(prefixCls, "-copy-disabled") : '', props.className), ref: ref, onClick: function (e) {
            var _a, _b, _c, _d, _e, _f;
            if (self === null || self === void 0 ? void 0 : self.disabled)
                return;
            e.stopPropagation();
            if ((_a = array.props) === null || _a === void 0 ? void 0 : _a.disabled)
                return;
            if (props.onClick) {
                props.onClick(e);
                if (e.defaultPrevented)
                    return;
            }
            var value = (0, shared_1.clone)((_b = array === null || array === void 0 ? void 0 : array.field) === null || _b === void 0 ? void 0 : _b.value[index]);
            var distIndex = index + 1;
            (_d = (_c = array.field) === null || _c === void 0 ? void 0 : _c.insert) === null || _d === void 0 ? void 0 : _d.call(_c, distIndex, value);
            (_f = (_e = array.props) === null || _e === void 0 ? void 0 : _e.onCopy) === null || _f === void 0 ? void 0 : _f.call(_e, distIndex);
        }, icon: (0, shared_1.isUndef)(props.icon) ? react_2.default.createElement(icons_1.CopyOutlined, null) : props.icon }), props.title || self.title));
});
var Remove = (0, react_2.forwardRef)(function (props, ref) {
    var _a;
    var index = useIndex(props.index);
    var self = (0, react_1.useField)();
    var array = useArray();
    var prefixCls = (0, __builtins__1.usePrefixCls)('formily-array-base');
    var _b = __read((0, style_1.default)(prefixCls), 2), wrapSSR = _b[0], hashId = _b[1];
    if (!array)
        return null;
    if (((_a = array.field) === null || _a === void 0 ? void 0 : _a.pattern) !== 'editable')
        return null;
    return wrapSSR(react_2.default.createElement(antd_1.Button, __assign({ type: "ghost" }, props, { style: __assign({ padding: '0 0 0 6px', width: 'auto', height: 'auto' }, props.style), disabled: self === null || self === void 0 ? void 0 : self.disabled, className: (0, classnames_1.default)("".concat(prefixCls, "-remove"), hashId, (self === null || self === void 0 ? void 0 : self.disabled) ? "".concat(prefixCls, "-remove-disabled") : '', props.className), ref: ref, onClick: function (e) {
            var _a, _b, _c, _d;
            if (self === null || self === void 0 ? void 0 : self.disabled)
                return;
            e.stopPropagation();
            if (props.onClick) {
                props.onClick(e);
                if (e.defaultPrevented)
                    return;
            }
            (_b = (_a = array.field) === null || _a === void 0 ? void 0 : _a.remove) === null || _b === void 0 ? void 0 : _b.call(_a, index);
            (_d = (_c = array.props) === null || _c === void 0 ? void 0 : _c.onRemove) === null || _d === void 0 ? void 0 : _d.call(_c, index);
        }, icon: (0, shared_1.isUndef)(props.icon) ? react_2.default.createElement(icons_1.DeleteOutlined, null) : props.icon }), props.title || self.title));
});
var MoveDown = (0, react_2.forwardRef)(function (props, ref) {
    var _a;
    var index = useIndex(props.index);
    var self = (0, react_1.useField)();
    var array = useArray();
    var prefixCls = (0, __builtins__1.usePrefixCls)('formily-array-base');
    if (!array)
        return null;
    if (((_a = array.field) === null || _a === void 0 ? void 0 : _a.pattern) !== 'editable')
        return null;
    return (react_2.default.createElement(antd_1.Button, __assign({ type: "ghost" }, props, { style: __assign({ padding: '0 0 0 6px', width: 'auto', height: 'auto' }, props.style), disabled: self === null || self === void 0 ? void 0 : self.disabled, className: (0, classnames_1.default)("".concat(prefixCls, "-move-down"), (self === null || self === void 0 ? void 0 : self.disabled) ? "".concat(prefixCls, "-move-down-disabled") : '', props.className), ref: ref, onClick: function (e) {
            var _a, _b, _c, _d;
            if (self === null || self === void 0 ? void 0 : self.disabled)
                return;
            e.stopPropagation();
            if (props.onClick) {
                props.onClick(e);
                if (e.defaultPrevented)
                    return;
            }
            (_b = (_a = array.field) === null || _a === void 0 ? void 0 : _a.moveDown) === null || _b === void 0 ? void 0 : _b.call(_a, index);
            (_d = (_c = array.props) === null || _c === void 0 ? void 0 : _c.onMoveDown) === null || _d === void 0 ? void 0 : _d.call(_c, index);
        }, icon: (0, shared_1.isUndef)(props.icon) ? react_2.default.createElement(icons_1.DownOutlined, null) : props.icon }), props.title || self.title));
});
var MoveUp = (0, react_2.forwardRef)(function (props, ref) {
    var _a;
    var index = useIndex(props.index);
    var self = (0, react_1.useField)();
    var array = useArray();
    var prefixCls = (0, __builtins__1.usePrefixCls)('formily-array-base');
    if (!array)
        return null;
    if (((_a = array.field) === null || _a === void 0 ? void 0 : _a.pattern) !== 'editable')
        return null;
    return (react_2.default.createElement(antd_1.Button, __assign({ type: "ghost" }, props, { style: __assign({ padding: '0 0 0 6px', width: 'auto', height: 'auto' }, props.style), disabled: self === null || self === void 0 ? void 0 : self.disabled, className: (0, classnames_1.default)("".concat(prefixCls, "-move-up"), (self === null || self === void 0 ? void 0 : self.disabled) ? "".concat(prefixCls, "-move-up-disabled") : '', props.className), ref: ref, onClick: function (e) {
            var _a, _b, _c;
            if (self === null || self === void 0 ? void 0 : self.disabled)
                return;
            e.stopPropagation();
            if (props.onClick) {
                props.onClick(e);
                if (e.defaultPrevented)
                    return;
            }
            (_a = array === null || array === void 0 ? void 0 : array.field) === null || _a === void 0 ? void 0 : _a.moveUp(index);
            (_c = (_b = array === null || array === void 0 ? void 0 : array.props) === null || _b === void 0 ? void 0 : _b.onMoveUp) === null || _c === void 0 ? void 0 : _c.call(_b, index);
        }, icon: (0, shared_1.isUndef)(props.icon) ? react_2.default.createElement(icons_1.UpOutlined, null) : props.icon }), props.title || self.title));
});
function mixin(target) {
    return Object.assign(target, {
        Index: Index,
        SortHandle: SortHandle,
        Addition: Addition,
        Copy: Copy,
        Remove: Remove,
        MoveDown: MoveDown,
        MoveUp: MoveUp,
        useArray: useArray,
        useIndex: useIndex,
        useRecord: useRecord,
    });
}
exports.ArrayBase = Object.assign(InternalArrayBase, {
    Item: Item,
    Index: Index,
    SortHandle: SortHandle,
    Addition: Addition,
    Copy: Copy,
    Remove: Remove,
    MoveDown: MoveDown,
    MoveUp: MoveUp,
    useArray: useArray,
    useIndex: useIndex,
    useRecord: useRecord,
    mixin: mixin,
});
exports.default = exports.ArrayBase;
