// @formily/antd-v5/lib/preview-text/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreviewText = exports.usePlaceholder = exports.Placeholder = exports.PlaceholderContext = void 0;
var react_1 = require("@formily/react");
var shared_1 = require("@formily/shared");
var antd_1 = require("antd");
var classnames_1 = __importDefault(require("classnames"));
var react_2 = __importStar(require("react"));
var __builtins__1 = require("../__builtins__");
var style_1 = __importDefault(require("./style"));
exports.PlaceholderContext = (0, react_2.createContext)('N/A');
exports.Placeholder = exports.PlaceholderContext.Provider;
var usePlaceholder = function (value) {
    var placeholder = (0, react_2.useContext)(exports.PlaceholderContext) || 'N/A';
    return (0, shared_1.isValid)(value) && value !== '' ? value : placeholder;
};
exports.usePlaceholder = usePlaceholder;
var Input = (0, react_1.observer)(function (props) {
    var prefixCls = (0, __builtins__1.usePrefixCls)('form-text', props);
    var _a = __read((0, style_1.default)(prefixCls), 2), wrapSSR = _a[0], hashId = _a[1];
    return wrapSSR(react_2.default.createElement(antd_1.Space, { className: (0, classnames_1.default)(prefixCls, hashId, props.className), style: props.style },
        props.addonBefore,
        props.prefix,
        (0, exports.usePlaceholder)(props.value),
        props.suffix,
        props.addonAfter));
});
var NumberPicker = (0, react_1.observer)(function (props) {
    var prefixCls = (0, __builtins__1.usePrefixCls)('form-text', props);
    var _a = __read((0, style_1.default)(prefixCls), 2), wrapSSR = _a[0], hashId = _a[1];
    return wrapSSR(react_2.default.createElement(antd_1.Space, { className: (0, classnames_1.default)(prefixCls, hashId, props.className), style: props.style },
        props.addonBefore,
        props.prefix,
        (0, exports.usePlaceholder)(props.formatter
            ? props.formatter(String(props.value), {
                userTyping: false,
                input: '',
            })
            : props.value),
        props['suffix'],
        props.addonAfter));
});
var Select = (0, react_1.observer)(function (props) {
    var _a, _b;
    var field = (0, react_1.useField)();
    var prefixCls = (0, __builtins__1.usePrefixCls)('form-text', props);
    var _c = __read((0, style_1.default)(prefixCls), 2), wrapSSR = _c[0], hashId = _c[1];
    var dataSource = ((_a = field === null || field === void 0 ? void 0 : field.dataSource) === null || _a === void 0 ? void 0 : _a.length)
        ? field.dataSource
        : ((_b = props === null || props === void 0 ? void 0 : props.options) === null || _b === void 0 ? void 0 : _b.length)
            ? props.options
            : [];
    var placeholder = (0, exports.usePlaceholder)();
    var getSelected = function () {
        var value = props.value;
        if (props.mode === 'multiple' || props.mode === 'tags') {
            if (props.labelInValue) {
                return (0, shared_1.isArr)(value) ? value : [];
            }
            else {
                return (0, shared_1.isArr)(value)
                    ? value.map(function (val) { return ({ label: val, value: val }); })
                    : [];
            }
        }
        else {
            if (props.labelInValue) {
                return (0, shared_1.isValid)(value) ? [value] : [];
            }
            else {
                return (0, shared_1.isValid)(value) ? [{ label: value, value: value }] : [];
            }
        }
    };
    var getLabel = function (target) {
        var _a, _b;
        var labelKey = ((_a = props.fieldNames) === null || _a === void 0 ? void 0 : _a.label) || 'label';
        return (((_b = dataSource === null || dataSource === void 0 ? void 0 : dataSource.find(function (item) {
            var _a;
            var valueKey = ((_a = props.fieldNames) === null || _a === void 0 ? void 0 : _a.value) || 'value';
            return item[valueKey] == (target === null || target === void 0 ? void 0 : target.value);
        })) === null || _b === void 0 ? void 0 : _b[labelKey]) ||
            target.label ||
            placeholder);
    };
    var getLabels = function () {
        var selected = getSelected();
        if (!selected.length)
            return placeholder;
        if (selected.length === 1)
            return getLabel(selected[0]);
        return selected.map(function (item, key) {
            return react_2.default.createElement(antd_1.Tag, { key: key }, getLabel(item));
        });
    };
    return wrapSSR(react_2.default.createElement("div", { className: (0, classnames_1.default)(prefixCls, hashId, props.className), style: props.style }, getLabels()));
});
var TreeSelect = (0, react_1.observer)(function (props) {
    var _a, _b;
    var field = (0, react_1.useField)();
    var placeholder = (0, exports.usePlaceholder)();
    var prefixCls = (0, __builtins__1.usePrefixCls)('form-text', props);
    var _c = __read((0, style_1.default)(prefixCls), 2), wrapSSR = _c[0], hashId = _c[1];
    var dataSource = ((_a = field === null || field === void 0 ? void 0 : field.dataSource) === null || _a === void 0 ? void 0 : _a.length)
        ? field.dataSource
        : ((_b = props === null || props === void 0 ? void 0 : props.treeData) === null || _b === void 0 ? void 0 : _b.length)
            ? props.treeData
            : [];
    var getSelected = function () {
        var value = props.value;
        if (props.multiple) {
            if (props.labelInValue) {
                return (0, shared_1.isArr)(value) ? value : [];
            }
            else {
                return (0, shared_1.isArr)(value)
                    ? value.map(function (val) { return ({ label: val, value: val }); })
                    : [];
            }
        }
        else {
            if (props.labelInValue) {
                return value ? [value] : [];
            }
            else {
                return value ? [{ label: value, value: value }] : [];
            }
        }
    };
    var findLabel = function (value, dataSource, treeNodeLabelProp) {
        var _a, _b;
        var valueKey = ((_a = props.fieldNames) === null || _a === void 0 ? void 0 : _a.value) || 'value';
        for (var i = 0; i < (dataSource === null || dataSource === void 0 ? void 0 : dataSource.length); i++) {
            var item = dataSource[i];
            if ((item === null || item === void 0 ? void 0 : item[valueKey]) === value) {
                return (_b = item === null || item === void 0 ? void 0 : item.label) !== null && _b !== void 0 ? _b : item[treeNodeLabelProp];
            }
            else {
                var childLabel = findLabel(value, item === null || item === void 0 ? void 0 : item.children, treeNodeLabelProp);
                if (childLabel)
                    return childLabel;
            }
        }
    };
    var getLabels = function () {
        var selected = getSelected();
        if (!(selected === null || selected === void 0 ? void 0 : selected.length))
            return react_2.default.createElement(antd_1.Tag, null, placeholder);
        return selected.map(function (_a, key) {
            var value = _a.value, label = _a.label;
            return (react_2.default.createElement(antd_1.Tag, { key: key }, findLabel(value, dataSource, props.treeNodeLabelProp) ||
                label ||
                placeholder));
        });
    };
    return wrapSSR(react_2.default.createElement("div", { className: (0, classnames_1.default)(prefixCls, hashId, props.className), style: props.style }, getLabels()));
});
var Cascader = (0, react_1.observer)(function (props) {
    var _a, _b;
    var field = (0, react_1.useField)();
    var placeholder = (0, exports.usePlaceholder)();
    var prefixCls = (0, __builtins__1.usePrefixCls)('form-text', props);
    var _c = __read((0, style_1.default)(prefixCls), 2), wrapSSR = _c[0], hashId = _c[1];
    var dataSource = ((_a = field === null || field === void 0 ? void 0 : field.dataSource) === null || _a === void 0 ? void 0 : _a.length)
        ? field.dataSource
        : ((_b = props === null || props === void 0 ? void 0 : props.options) === null || _b === void 0 ? void 0 : _b.length)
            ? props.options
            : [];
    var findSelectedItem = function (items, val) {
        return items.find(function (item) { return item.value == val; });
    };
    var findSelectedItems = function (sources, selectedValues) {
        return selectedValues.map(function (value) {
            var result = [];
            var items = sources;
            value.forEach(function (val) {
                var _a, _b;
                var selectedItem = findSelectedItem(items, val);
                result.push({
                    label: (_a = selectedItem === null || selectedItem === void 0 ? void 0 : selectedItem.label) !== null && _a !== void 0 ? _a : '',
                    value: selectedItem === null || selectedItem === void 0 ? void 0 : selectedItem.value,
                });
                items = (_b = selectedItem === null || selectedItem === void 0 ? void 0 : selectedItem.children) !== null && _b !== void 0 ? _b : [];
            });
            return result;
        });
    };
    var getSelected = function () {
        var val = (0, shared_1.toArr)(props.value);
        // unified conversion to multi selection mode
        return props.multiple ? val : [val];
    };
    var getLabels = function () {
        var selected = getSelected();
        var values = findSelectedItems(dataSource, selected);
        var labels = values
            .map(function (val) {
            return val.map(function (item) { return item.label; }).join('/');
        })
            .join(' ');
        return labels || placeholder;
    };
    return wrapSSR(react_2.default.createElement("div", { className: (0, classnames_1.default)(prefixCls, hashId, props.className), style: props.style }, getLabels()));
});
var DatePicker = function (props) {
    var placeholder = (0, exports.usePlaceholder)();
    var prefixCls = (0, __builtins__1.usePrefixCls)('form-text', props);
    var getLabels = function () {
        var labels = (0, __builtins__1.formatDayjsValue)(props.value, props.format, placeholder);
        return (0, shared_1.isArr)(labels) ? labels.join('~') : labels;
    };
    return react_2.default.createElement("div", { className: (0, classnames_1.default)(prefixCls, props.className) }, getLabels());
};
var DateRangePicker = function (props) {
    var placeholder = (0, exports.usePlaceholder)();
    var prefixCls = (0, __builtins__1.usePrefixCls)('form-text', props);
    var getLabels = function () {
        var labels = (0, __builtins__1.formatDayjsValue)(props.value, props.format, placeholder);
        return (0, shared_1.isArr)(labels) ? labels.join('~') : labels;
    };
    return (react_2.default.createElement("div", { className: (0, classnames_1.default)(prefixCls, props.className), style: props.style }, getLabels()));
};
var TimePicker = function (props) {
    var placeholder = (0, exports.usePlaceholder)();
    var prefixCls = (0, __builtins__1.usePrefixCls)('form-text', props);
    var getLabels = function () {
        var labels = (0, __builtins__1.formatDayjsValue)(props.value, props.format, placeholder);
        return (0, shared_1.isArr)(labels) ? labels.join('~') : labels;
    };
    return (react_2.default.createElement("div", { className: (0, classnames_1.default)(prefixCls, props.className), style: props.style }, getLabels()));
};
var TimeRangePicker = function (props) {
    var placeholder = (0, exports.usePlaceholder)();
    var prefixCls = (0, __builtins__1.usePrefixCls)('form-text', props);
    var getLabels = function () {
        var labels = (0, __builtins__1.formatDayjsValue)(props.value, props.format, placeholder);
        return (0, shared_1.isArr)(labels) ? labels.join('~') : labels;
    };
    return (react_2.default.createElement("div", { className: (0, classnames_1.default)(prefixCls, props.className), style: props.style }, getLabels()));
};
var InternalPreviewText = (0, react_1.observer)(function (props) {
    var prefixCls = (0, __builtins__1.usePrefixCls)('form-text', props);
    return (react_2.default.createElement("div", { className: (0, classnames_1.default)(prefixCls, props.className), style: props.style }, (0, exports.usePlaceholder)(props.value)));
});
exports.PreviewText = Object.assign(InternalPreviewText, {
    Input: Input,
    Select: Select,
    TreeSelect: TreeSelect,
    Cascader: Cascader,
    DatePicker: DatePicker,
    DateRangePicker: DateRangePicker,
    TimePicker: TimePicker,
    TimeRangePicker: TimeRangePicker,
    Placeholder: exports.Placeholder,
    usePlaceholder: exports.usePlaceholder,
    NumberPicker: NumberPicker,
});
exports.default = exports.PreviewText;
