// @formily/antd-v5/esm/__builtins__/sort.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { DndContext } from '@dnd-kit/core';
import { SortableContext, useSortable, verticalListSortingStrategy, } from '@dnd-kit/sortable';
import React, { createContext, useContext, useMemo } from 'react';
export function SortableContainer(Component) {
    return function (_a) {
        var list = _a.list, _b = _a.start, start = _b === void 0 ? 0 : _b, accessibility = _a.accessibility, onSortStart = _a.onSortStart, onSortEnd = _a.onSortEnd, props = __rest(_a, ["list", "start", "accessibility", "onSortStart", "onSortEnd"]);
        var _onSortEnd = function (event) {
            var active = event.active, over = event.over;
            if (!over)
                return;
            var oldIndex = +active.id - 1;
            var newIndex = +over.id - 1;
            onSortEnd === null || onSortEnd === void 0 ? void 0 : onSortEnd({
                oldIndex: oldIndex,
                newIndex: newIndex,
            });
        };
        return (React.createElement(DndContext, { accessibility: accessibility, onDragStart: onSortStart, onDragEnd: _onSortEnd },
            React.createElement(SortableContext, { items: list.map(function (_, index) { return index + start + 1; }), strategy: verticalListSortingStrategy },
                React.createElement(Component, __assign({}, props), props.children))));
    };
}
export var useSortableItem = function () {
    return useContext(SortableItemContext);
};
export var SortableItemContext = createContext({});
export function SortableElement(Component) {
    return function (_a) {
        var _b = _a.index, index = _b === void 0 ? 0 : _b, lockAxis = _a.lockAxis, props = __rest(_a, ["index", "lockAxis"]);
        var sortable = useSortable({
            id: index + 1,
        });
        var setNodeRef = sortable.setNodeRef, transform = sortable.transform, transition = sortable.transition, isDragging = sortable.isDragging;
        if (transform) {
            switch (lockAxis) {
                case 'x':
                    transform.y = 0;
                    break;
                case 'y':
                    transform.x = 0;
                    break;
                default:
                    break;
            }
        }
        var style = useMemo(function () {
            var zIndex = transform ? 1 : 'none';
            var position = transform ? 'relative' : 'unset';
            var itemStyle = {
                position: position,
                touchAction: 'none',
                zIndex: zIndex,
                transform: transform
                    ? "translate3d(".concat(transform.x, "px, ").concat(transform.y, "px, 0)")
                    : 'none',
                transition: "".concat(transform ? 'all 200ms ease' : ''),
            };
            var dragStyle = {
                transition: transition,
                opacity: '0.8',
                transform: "translate3d(".concat((transform === null || transform === void 0 ? void 0 : transform.x) || 0, "px, ").concat((transform === null || transform === void 0 ? void 0 : transform.y) || 0, "px, 0)"),
            };
            var computedStyle = isDragging
                ? __assign(__assign(__assign({}, itemStyle), dragStyle), props.style) : __assign(__assign({}, itemStyle), props.style);
            return computedStyle;
        }, [isDragging, transform, transition, props.style]);
        return (React.createElement(SortableItemContext.Provider, { value: sortable }, Component(__assign(__assign({}, props), { style: style, ref: setNodeRef }))));
    };
}
export function SortableHandle(Component) {
    return function (props) {
        var _a = useSortableItem(), attributes = _a.attributes, listeners = _a.listeners;
        return React.createElement(Component, __assign({}, props, attributes, listeners));
    };
}
