// @formily/antd-v5/esm/array-items/style.js
import { genStyleHook } from './../__builtins__';
export default genStyleHook('array-items', function (token) {
    var _a, _b, _c, _d, _e, _f;
    var componentCls = token.componentCls, antCls = token.antCls, colorBgBase = token.colorBgBase, fontSizeSM = token.fontSizeSM, colorBorder = token.colorBorder, colorText = token.colorText;
    var itemCls = "".concat(antCls, "-formily-item");
    return _a = {},
        _a[componentCls] = (_b = {},
            _b["".concat(componentCls, "-item-inner")] = {
                visibility: 'visible',
            },
            _b['&-item'] = {
                zIndex: 100000,
            },
            _b['&-card'] = (_c = {
                    display: 'flex',
                    border: "1px solid ".concat(colorBorder),
                    marginBottom: 10,
                    padding: '3px 6px',
                    background: colorBgBase,
                    justifyContent: 'space-between',
                    color: colorText
                },
                _c["".concat(itemCls, ":not(").concat(itemCls, "-feedback-layout-popover)")] = (_d = {
                        marginBottom: '0 !important'
                    },
                    _d["".concat(itemCls, "-help")] = {
                        position: 'absolute',
                        fontSize: fontSizeSM,
                        top: '100%',
                        background: colorBgBase,
                        width: '100%',
                        marginTop: 3,
                        padding: 3,
                        zIndex: 1,
                        borderRadius: 3,
                        boxShadow: "0 0 10px ".concat(colorBorder),
                    },
                    _d),
                _c),
            _b['&-divide'] = (_e = {
                    display: 'flex',
                    borderBottom: "1px solid ".concat(colorBorder),
                    padding: '10px 0',
                    justifyContent: 'space-between'
                },
                _e["".concat(itemCls, ":not(").concat(itemCls, "-feedback-layout-popover)")] = (_f = {
                        marginBottom: '0 !important'
                    },
                    _f["".concat(itemCls, "-help")] = {
                        position: 'absolute',
                        fontSize: fontSizeSM,
                        top: '100%',
                        background: colorBgBase,
                        width: '100%',
                        marginTop: 3,
                        padding: 3,
                        zIndex: 1,
                        borderRadius: 3,
                        boxShadow: "0 0 10px ".concat(colorBorder),
                    },
                    _f),
                _e),
            _b),
        _a;
});
