// @formily/antd-v5/esm/array-table/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
import { observer, RecursionField, useField, useFieldSchema, } from '@formily/react';
import { isArr, isBool, isFn } from '@formily/shared';
import { Badge, Pagination, Select, Space, Table, } from 'antd';
import cls from 'classnames';
import React, { createContext, forwardRef, Fragment, useContext, useEffect, useRef, useState, } from 'react';
import { ArrayBase } from '../array-base';
import { SortableContainer, SortableElement, usePrefixCls, } from '../__builtins__';
import useStyle from './style';
var SortableRow = SortableElement(function (props) { return React.createElement("tr", __assign({}, props)); });
var SortableBodyRaw = SortableContainer(function (_a) {
    var tbodyRef = _a.tbodyRef, props = __rest(_a, ["tbodyRef"]);
    return React.createElement("tbody", __assign({}, props, { ref: tbodyRef }));
});
var SortableBody = forwardRef(function (props, ref) { return React.createElement(SortableBodyRaw, __assign({}, props, { tbodyRef: ref })); });
var isColumnComponent = function (schema) {
    var _a;
    return ((_a = schema['x-component']) === null || _a === void 0 ? void 0 : _a.indexOf('Column')) > -1;
};
var isOperationsComponent = function (schema) {
    var _a;
    return ((_a = schema['x-component']) === null || _a === void 0 ? void 0 : _a.indexOf('Operations')) > -1;
};
var isAdditionComponent = function (schema) {
    var _a;
    return ((_a = schema['x-component']) === null || _a === void 0 ? void 0 : _a.indexOf('Addition')) > -1;
};
var useArrayTableSources = function () {
    var arrayField = useField();
    var schema = useFieldSchema();
    var parseSources = function (schema) {
        var _a, _b, _c;
        if (isColumnComponent(schema) ||
            isOperationsComponent(schema) ||
            isAdditionComponent(schema)) {
            if (!((_a = schema['x-component-props']) === null || _a === void 0 ? void 0 : _a['dataIndex']) && !schema['name'])
                return [];
            var name_1 = ((_b = schema['x-component-props']) === null || _b === void 0 ? void 0 : _b['dataIndex']) || schema['name'];
            var field = arrayField.query(arrayField.address.concat(name_1)).take();
            var columnProps = ((_c = field === null || field === void 0 ? void 0 : field.component) === null || _c === void 0 ? void 0 : _c[1]) || schema['x-component-props'] || {};
            var display = (field === null || field === void 0 ? void 0 : field.display) || schema['x-display'];
            return [
                {
                    name: name_1,
                    display: display,
                    field: field,
                    schema: schema,
                    columnProps: columnProps,
                },
            ];
        }
        else if (schema.properties) {
            return schema.reduceProperties(function (buf, schema) {
                return buf.concat(parseSources(schema));
            }, []);
        }
        return [];
    };
    var parseArrayItems = function (schema) {
        if (!schema)
            return [];
        var sources = [];
        var items = isArr(schema) ? schema : [schema];
        return items.reduce(function (columns, schema) {
            var item = parseSources(schema);
            if (item) {
                return columns.concat(item);
            }
            return columns;
        }, sources);
    };
    if (!schema)
        throw new Error('can not found schema object');
    return parseArrayItems(schema.items);
};
var indexKey = '__DO_NOT_USE_THIS_PROPERTY_index__';
var useArrayTableColumns = function (dataSource, field, sources) {
    return sources.reduce(function (buf, _a, key) {
        var name = _a.name, columnProps = _a.columnProps, schema = _a.schema, display = _a.display;
        if (display !== 'visible')
            return buf;
        if (!isColumnComponent(schema))
            return buf;
        return buf.concat(__assign(__assign({}, columnProps), { key: key, dataIndex: name, render: function (value, record) {
                var index = record[indexKey];
                var children = (React.createElement(ArrayBase.Item, { index: index, record: function () { var _a; return (_a = field === null || field === void 0 ? void 0 : field.value) === null || _a === void 0 ? void 0 : _a[index]; } },
                    React.createElement(RecursionField, { schema: schema, name: index, onlyRenderProperties: true })));
                return index > -1 ? children : null;
            } }));
    }, []);
};
var useAddition = function () {
    var schema = useFieldSchema();
    return schema.reduceProperties(function (addition, schema, key) {
        if (isAdditionComponent(schema)) {
            return React.createElement(RecursionField, { schema: schema, name: key });
        }
        return addition;
    }, null);
};
var schedulerRequest = {
    request: null,
};
var StatusSelect = observer(function (props) {
    var _a;
    var field = useField();
    var prefixCls = usePrefixCls('formily-array-table');
    var _b = __read(useStyle(prefixCls), 2), wrapSSR = _b[0], hashId = _b[1];
    var errors = field.errors;
    var parseIndex = function (address) {
        var _a;
        return Number((_a = address === null || address === void 0 ? void 0 : address.slice(address.indexOf(field.address.toString())).match(/(\d+)/)) === null || _a === void 0 ? void 0 : _a[1]);
    };
    var options = (_a = props.options) === null || _a === void 0 ? void 0 : _a.map(function (_a) {
        var label = _a.label, value = _a.value;
        var val = Number(value);
        var hasError = errors.some(function (_a) {
            var _b;
            var address = _a.address;
            var currentIndex = parseIndex(address);
            var startIndex = props.pageSize ? (val - 1) * props.pageSize : 0;
            var endIndex = props.pageSize
                ? val * props.pageSize
                : ((_b = props.options) === null || _b === void 0 ? void 0 : _b.length) || 0;
            return currentIndex >= startIndex && currentIndex <= endIndex;
        });
        return {
            label: hasError ? React.createElement(Badge, { dot: true }, label) : label,
            value: value,
        };
    });
    var width = String(options === null || options === void 0 ? void 0 : options.length).length * 15;
    return wrapSSR(React.createElement(Select, { value: props.value, onChange: props.onChange, options: options, virtual: true, style: {
            width: width < 60 ? 60 : width,
        }, className: cls("".concat(prefixCls, "-status-select"), hashId, {
            'has-error': errors === null || errors === void 0 ? void 0 : errors.length,
        }) }));
}, {
    scheduler: function (update) {
        clearTimeout(schedulerRequest.request);
        schedulerRequest.request = setTimeout(function () {
            update();
        }, 100);
    },
});
var PaginationContext = createContext({});
var usePagination = function () {
    return useContext(PaginationContext);
};
var ArrayTablePagination = function (props) {
    var _a;
    var _b = __read(useState(1), 2), current = _b[0], setCurrent = _b[1];
    var prefixCls = usePrefixCls('formily-array-table');
    var _c = __read(useStyle(prefixCls), 2), wrapSSR = _c[0], hashId = _c[1];
    var _d = __read(useState(props.pageSize || 10), 2), pageSize = _d[0], setPageSize = _d[1];
    var size = props.size || 'default';
    var dataSource = props.dataSource || [];
    var showPagination = props.showPagination;
    var startIndex = (current - 1) * pageSize;
    var endIndex = startIndex + pageSize - 1;
    var total = (dataSource === null || dataSource === void 0 ? void 0 : dataSource.length) || 0;
    var totalPage = Math.ceil(total / pageSize);
    var pages = Array.from(new Array(totalPage)).map(function (_, index) {
        var page = index + 1;
        return {
            label: page,
            value: page,
        };
    });
    var handleChange = function (current) {
        setCurrent(current);
    };
    var handleSizeChange = function (_, size) {
        setPageSize(size);
    };
    useEffect(function () {
        setPageSize(props.pageSize || 10);
    }, [props.pageSize]);
    useEffect(function () {
        if (totalPage > 0 && totalPage < current) {
            handleChange(totalPage);
        }
    }, [totalPage, current]);
    var renderPagination = function () {
        if (!showPagination || totalPage < 1)
            return;
        return (React.createElement("div", { className: cls("".concat(prefixCls, "-pagination"), hashId) },
            React.createElement(Space, null,
                React.createElement(StatusSelect, { value: current, pageSize: pageSize, onChange: handleChange, options: pages, notFoundContent: false }),
                React.createElement(Pagination, __assign({}, props, { pageSize: pageSize, current: current, total: dataSource.length, size: size, showSizeChanger: true, onShowSizeChange: handleSizeChange, onChange: handleChange })))));
    };
    return wrapSSR(React.createElement(Fragment, null,
        React.createElement(PaginationContext.Provider, { value: { totalPage: totalPage, pageSize: pageSize, startIndex: startIndex, changePage: handleChange } }, (_a = props.children) === null || _a === void 0 ? void 0 : _a.call(props, showPagination
            ? dataSource === null || dataSource === void 0 ? void 0 : dataSource.slice(startIndex, endIndex + 1)
            : dataSource, renderPagination(), {
            startIndex: startIndex,
        }))));
};
var WrapperComp = function (props) {
    var ref = useRef(null);
    var field = useField();
    var prefixCls = usePrefixCls('formily-array-table');
    var startIndex = useContext(PaginationContext).startIndex;
    var dataSource = Array.isArray(field.value) ? field.value.slice() : [];
    var addTdStyles = function (id) {
        var _a;
        var node = (_a = ref.current) === null || _a === void 0 ? void 0 : _a.querySelector(".".concat(prefixCls, "-row-").concat(id));
        var helper = document.body.querySelector(".".concat(prefixCls, "-sort-helper"));
        if (!helper)
            return;
        var tds = node === null || node === void 0 ? void 0 : node.querySelectorAll('td');
        if (!tds)
            return;
        requestAnimationFrame(function () {
            helper.querySelectorAll('td').forEach(function (td, index) {
                if (tds[index]) {
                    td.style.width = getComputedStyle(tds[index]).width;
                }
            });
        });
    };
    return (React.createElement(SortableBody, __assign({ ref: ref }, props, { start: startIndex, list: dataSource.slice(), accessibility: {
            container: ref.current || undefined,
        }, onSortStart: function (event) {
            addTdStyles(event.active.id);
        }, onSortEnd: function (_a) {
            var oldIndex = _a.oldIndex, newIndex = _a.newIndex;
            field.move(oldIndex, newIndex);
        }, className: cls("".concat(prefixCls, "-sort-helper"), props.className) })));
};
var RowComp = function (props) {
    var prefixCls = usePrefixCls('formily-array-table');
    var index = props['data-row-key'] || 0;
    return (React.createElement(SortableRow, __assign({ lockAxis: "y" }, props, { index: index, className: cls(props.className, "".concat(prefixCls, "-row-").concat(index + 1)) })));
};
var InternalArrayTable = observer(function (props) {
    var ref = useRef(null);
    var field = useField();
    var prefixCls = usePrefixCls('formily-array-table');
    var _a = __read(useStyle(prefixCls), 2), wrapSSR = _a[0], hashId = _a[1];
    var dataSource = Array.isArray(field.value) ? field.value.slice() : [];
    dataSource.forEach(function (item, index) {
        item[indexKey] = index;
    });
    var sources = useArrayTableSources();
    var columns = useArrayTableColumns(dataSource, field, sources);
    var pagination = isBool(props.pagination) ? {} : props.pagination;
    var showPagination = isBool(props.pagination) ? props.pagination : true;
    var addition = useAddition();
    var defaultRowKey = function (record) {
        return record[indexKey];
    };
    return wrapSSR(React.createElement(ArrayTablePagination, __assign({}, pagination, { dataSource: dataSource, showPagination: showPagination }), function (dataSource, pager, _a) {
        var startIndex = _a.startIndex;
        return (React.createElement("div", { ref: ref, className: cls(prefixCls, hashId) },
            React.createElement(ArrayBase, null,
                React.createElement(Table, __assign({ size: "small", bordered: true, rowKey: defaultRowKey }, props, { onChange: function () { }, pagination: false, columns: columns, dataSource: dataSource, components: {
                        body: {
                            wrapper: WrapperComp,
                            row: RowComp,
                        },
                    } })),
                React.createElement("div", { style: { marginTop: 5, marginBottom: 5 } }, pager),
                sources.map(function (column, key) {
                    //专门用来承接对Column的状态管理
                    if (!isColumnComponent(column.schema))
                        return;
                    return React.createElement(RecursionField, {
                        name: column.name,
                        schema: column.schema,
                        onlyRenderSelf: true,
                        key: key,
                    });
                }),
                addition)));
    }));
});
var Column = function () {
    return React.createElement(Fragment, null);
};
var Addition = function (props) {
    var array = ArrayBase.useArray();
    var _a = usePagination(), _b = _a.totalPage, totalPage = _b === void 0 ? 0 : _b, _c = _a.pageSize, pageSize = _c === void 0 ? 10 : _c, changePage = _a.changePage;
    return (React.createElement(ArrayBase.Addition, __assign({}, props, { onClick: function (e) {
            var _a, _b;
            // 如果添加数据后将超过当前页，则自动切换到下一页
            var total = ((_a = array === null || array === void 0 ? void 0 : array.field) === null || _a === void 0 ? void 0 : _a.value.length) || 0;
            if (total === totalPage * pageSize + 1 && isFn(changePage)) {
                changePage(totalPage + 1);
            }
            (_b = props.onClick) === null || _b === void 0 ? void 0 : _b.call(props, e);
        } })));
};
export var ArrayTable = Object.assign(ArrayBase.mixin(InternalArrayTable), {
    Column: Column,
    Addition: Addition,
});
ArrayTable.displayName = 'ArrayTable';
export default ArrayTable;
