// @formily/antd-v5/esm/array-table/style.js
import { genStyleHook } from '../__builtins__';
export default genStyleHook('array-table', function (token) {
    var _a, _b, _c, _d, _e, _f;
    var componentCls = token.componentCls, antCls = token.antCls, colorErrorBorder = token.colorErrorBorder, fontSizeSM = token.fontSizeSM, colorBgBase = token.colorBgBase, colorBorder = token.colorBorder, colorBgContainer = token.colorBgContainer;
    var itemCls = "".concat(antCls, "-formily-item");
    return _a = {},
        _a[componentCls] = (_b = {},
            _b["".concat(componentCls, "-pagination")] = (_c = {
                    display: 'flex',
                    justifyContent: 'center'
                },
                _c["".concat(componentCls, "-status-select.has-error")] = (_d = {},
                    _d["".concat(antCls, "-select-selector")] = {
                        borderColor: "".concat(colorErrorBorder, " !important"),
                    },
                    _d),
                _c),
            _b["".concat(antCls, "-table")] = {
                td: (_e = {
                        visibility: 'visible'
                    },
                    _e["".concat(itemCls, ":not(").concat(itemCls, "-feedback-layout-popover)")] = (_f = {
                            marginBottom: '0 !important'
                        },
                        _f["".concat(itemCls, "-help")] = {
                            position: 'absolute',
                            fontSize: fontSizeSM,
                            top: '100%',
                            background: colorBgBase,
                            width: '100%',
                            marginTop: 3,
                            padding: 3,
                            zIndex: 1,
                            borderRadius: 3,
                            boxShadow: "0 0 10px ".concat(colorBorder),
                            animation: 'none',
                            transform: 'translateY(0)',
                            opacity: 1,
                        },
                        _f),
                    _e),
                table: {
                    overflow: 'hidden',
                },
            },
            _b["".concat(componentCls, "-sort-helper")] = {
                background: colorBgContainer,
                border: "1px solid ".concat(colorBorder),
                zIndex: 10,
            },
            _b),
        _a;
});
