// @formily/antd-v5/lib/form-item/style/animation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAnimationStyle = void 0;
var cssinjs_1 = require("@ant-design/cssinjs");
var antShowHelpIn = new cssinjs_1.Keyframes('antShowHelpIn', {
    '0%': {
        transform: 'translateY(-5px)',
        opacity: 0,
    },
    to: {
        transform: ' translateY(0)',
        opacity: 1,
    },
});
var getAnimationStyle = function (token) {
    var _a;
    var componentCls = token.componentCls;
    var helpCls = "".concat(componentCls, "-help");
    return _a = {},
        _a[helpCls] = {
            '&-appear, &-enter': {
                animationDuration: '0.3s',
                animationFillMode: 'both',
                animationPlayState: 'paused',
                opacity: 0,
                animationTimingFunction: 'cubic-bezier(0.645, 0.045, 0.355, 1)',
                '&-active': {
                    animationPlayState: 'running',
                    animationName: antShowHelpIn,
                },
            },
        },
        _a;
};
exports.getAnimationStyle = getAnimationStyle;
