// @formily/antd-v5/esm/form-tab/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { observer, RecursionField, useField, useFieldSchema, } from '@formily/react';
import { markRaw, model } from '@formily/reactive';
import { Badge, Tabs } from 'antd';
import cls from 'classnames';
import React, { Fragment, useMemo } from 'react';
import { usePrefixCls } from '../__builtins__';
var useTabs = function () {
    var tabsField = useField();
    var schema = useFieldSchema();
    var tabs = [];
    schema.mapProperties(function (schema, name) {
        var _a, _b, _c;
        var field = tabsField.query(tabsField.address.concat(name)).take();
        if ((field === null || field === void 0 ? void 0 : field.display) === 'none' || (field === null || field === void 0 ? void 0 : field.display) === 'hidden')
            return;
        if (((_a = schema['x-component']) === null || _a === void 0 ? void 0 : _a.indexOf('TabPane')) > -1) {
            var key = ((_b = field === null || field === void 0 ? void 0 : field.componentProps) === null || _b === void 0 ? void 0 : _b.key) || ((_c = schema === null || schema === void 0 ? void 0 : schema['x-component-props']) === null || _c === void 0 ? void 0 : _c.key) || name;
            tabs.push({
                name: name,
                props: __assign(__assign(__assign({}, schema === null || schema === void 0 ? void 0 : schema['x-component-props']), field === null || field === void 0 ? void 0 : field.componentProps), { key: key }),
                schema: schema,
            });
        }
    });
    return tabs;
};
var FeedbackBadge = observer(function (props) {
    var field = useField();
    var errors = field.form.queryFeedbacks({
        type: 'error',
        address: "".concat(field.address.concat(props.name), ".*"),
    });
    if (errors.length) {
        return (React.createElement(Badge, { size: "small", className: "errors-badge", count: errors.length }, props.tab));
    }
    return React.createElement(Fragment, null, props.tab);
});
var createFormTab = function (defaultActiveKey) {
    var formTab = model({
        activeKey: defaultActiveKey,
        setActiveKey: function (key) {
            formTab.activeKey = key;
        },
    });
    return markRaw(formTab);
};
export var FormTab = observer(function (_a) {
    var formTab = _a.formTab, props = __rest(_a, ["formTab"]);
    var tabs = useTabs();
    var _formTab = useMemo(function () {
        return formTab ? formTab : createFormTab();
    }, []);
    var prefixCls = usePrefixCls('formily-tab', props);
    var activeKey = props.activeKey || (_formTab === null || _formTab === void 0 ? void 0 : _formTab.activeKey);
    return (React.createElement(Tabs, __assign({}, props, { className: cls(prefixCls, props.className), activeKey: activeKey, onChange: function (key) {
            var _a, _b;
            (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, key);
            (_b = _formTab === null || _formTab === void 0 ? void 0 : _formTab.setActiveKey) === null || _b === void 0 ? void 0 : _b.call(_formTab, key);
        }, items: tabs.map(function (_a, key) {
            var props = _a.props, schema = _a.schema, name = _a.name;
            return __assign(__assign({ key: key, label: React.createElement(FeedbackBadge, { name: name, tab: props.tab }) }, props), { children: React.createElement(RecursionField, { schema: schema, name: name }) });
        }) })));
});
var TabPane = function (_a) {
    var children = _a.children;
    return React.createElement(Fragment, null, children);
};
FormTab.TabPane = TabPane;
FormTab.createFormTab = createFormTab;
export default FormTab;
